### PR TITLE
Refactoring of module.

### PR DESCRIPTION
**Pull request by crayfishx**
## Summary

I've done a fair bit of refactoring to this module that I'd like to merge upstream in the hope of a next major release.  It's quite a large change so I've detailed as much as I can below.....
### Replacement of `config_hash` (API change)

This PR removes the generic `config_hash` attribute and replaces it with more fine grained controls by adding the following attributes to the `vault` class
- `backend` (Hash)
- `listener` (Hash)
- `ha_backend` (Hash, optional)
- `telemetry` (Hash, optional)
- `disable_mlock` (Boolean, optional)
- `disable_cache` (Boolean, optional)
- `default_lease_ttl` (String, optional)
- `max_lease_ttl` (String, optional)
- `disable_mlock` (Boolean, optional)

This has many benefits, especially when used with Hiera...  For example to disable mlock in my dev environment I would have to re-create the entire hash at the level in the hierarchy, with this change you can just set `vault::disable_mlock: true` at the appropriate place.   This also improves validation, and makes the module more understandable.

This is an API breaking change.  If you would prefer to maintain a backwardly compatible version of the module and deprecate the `config_hash` variable over the course of several releases, this would be fairly easy to implement.   I could put `config_hash` back in with a default of `undef` and put a conditional around the building of `vault::config::config_hash` to use this value if it's supplied instead of the new attributes, and document the attribute as deprecated.    I have not done this as I don't feel a 0.x module needs to worry too much about deprecating things over multiple releases - but I'm happy to add it if requested.
### Allow for package installations (#36)

As this was quite a major refactor I decided to pull in the PR from @miso321 adding package installations into this PR since there have been no objections to it, and it's currently unmerged.

Closes #36 
### Replaced `nanliu/staging` for `puppet/archive`

This is a modernisation, `nanliu/staging` is currently not under active feature development, the advised replacement for it is `puppet/archive`.
### JSON file now has line breaks.

The `vault_sorted_json` function now gives a prettier output with line breaks and indenting by using `JSON.pretty_generate`
### Other changes
- Changed the `scope.lookupvar()` notation in templates to `scope[]`, this is more best practice
- Updated spec tests (rspec passes for me)
- `puppetlabs/stdlib` dependancy set to 4.3 (for the `delete_undef_values` function
- Added documentation to the README for all class parameters

### CODE REVIEW

**Comment by jsok** [comment]
Thanks for all this work, much appreciated! Will review asap.

File: README.md [hunk 1]
```diff
@@ -21,55 +21,144 @@ This module is currently only tested on:
...
+* `service_provider`: Customise the name of the system service provider; this also controls the init configuration files that are installed.
+
+* `service_options`: Extra argument to pass to `vault server`, as per: `vault server --help`
+
+* `num_procs`: Sets the GOMAXPROCS environment variable, to determine how many CPUs Vault can use. The official Vault Terraform install.sh script sets this to the output of ``nprocs``, with the comment, "Make sure to use all our CPUs, because Vault can block a scheduler thread". Default: number of CPUs on the system, retrieved from the ``processorcount`` fact.
+
+
+### Configuration parameters
+
+By default, with no parameters the module will configure vault with some sensible defaults to get you running, the following parameters may be specified to configure Vault.  Please see [The official documentation](https://www.vaultproject.io/docs/config/) for further details of acceptable parameter values.
```
Comment (jsok) [inline README.md:77]: My original reasoning for making `config_hash` one big hash was to reduce the maintenance burden of having to track new and removed options from vault. It also means a user can use an older or RC release of vault where the configuration is different to what this module supports.

Comment (jsok) [reply README.md:77]: What about including a `config_extras` or similar option so users can provide a config that this module doesn't support?

File: manifests/install.pp [hunk 2]
```diff
@@ -3,18 +3,42 @@
...
-    creates => $vault_bin,
-  } ~>
+  case $::vault::install_method {
+      'archive': {
+        if $::vault::manage_download_dir {
+          file { $::vault::download_dir:
+            ensure => directory,
+          }
+        }
+        archive { "${::vault::download_dir}/${::vault::download_filename}":
```
Comment (jsok) [inline manifests/install.pp:13]: 👍 been on my list of TODOS. Pity it's so much more verbose than staging.

File: metadata.json [hunk 3]
```diff
@@ -10,11 +10,11 @@
   "dependencies": [
     {
       "name": "puppetlabs-stdlib",
-      "version_requirement": ">= 4.2.0"
+      "version_requirement": ">= 4.3.0"
```
Comment (jsok) [inline metadata.json]: Any new functionality that this PR requires? Or just to stay up to date?

File: README.md (hunk 1 above)
Comment (crayfishx) [reply README.md:77]: The maintenance burden shouldn't be too high, hashicorp are normally reasonably good about sustainable config, but sure, I can put an `extra_config` hash attribute in to catch any exceptions or future new config that the module doesn't support yet

File: metadata.json (hunk 3 above)
Comment (crayfishx) [reply metadata.json]: Oops - my mistake....

https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/CHANGELOG.md#2014-05-08---release---420

I thought `delete_undef_values` was released in 4.3, it was actually 4.2 so I'll revert this change.

**Comment by crayfishx** [comment]
@jsok I've added `extra_config` as an optional hash and reverted stdlib back to 4.2

**Review by jsok — APPROVED** [review]
LGTM, will merge now.

**Pull request merged**